runtime/compiler/aarch64/codegen/J9TreeEvaluator.cpp

```diff
@@ -682,12 +684,470 @@ J9::ARM64::TreeEvaluator::asynccheckEvaluator(TR::Node *node, TR::CodeGenerator
+   {
+   TR::Compilation *comp = cg->comp();
+
+   // Compare the instance class depth to the cast class depth. If the instance class depth is less than or equal to
+   // to the cast class depth then the cast class cannot be a superclass of the instance class.
+   //
+   TR::Register *instanceClassDepthReg = srm->findOrCreateScratchRegister();
+   TR::Register *castClassDepthReg = NULL;
+   static_assert(J9AccClassDepthMask == 0xffff, "J9_JAVA_CLASS_DEPTH_MASK must be 0xffff");
+   // load lower 16bit of classDepthAndFlags
+   generateTrg1MemInstruction(cg, TR::InstOpCode::ldrhimm, node, instanceClassDepthReg,
+                              new (cg->trHeapMemory()) TR::MemoryReference(instanceClassReg, offsetof(J9Class, classDepthAndFlags), cg));
+   if (constantIsUnsignedImm12(castClassDepth))
+      {
+      generateCompareImmInstruction(cg, node, instanceClassDepthReg, castClassDepth);
+      }
+   else
+      {
+      castClassDepthReg = srm->findOrCreateScratchRegister();
+      loadConstant32(cg, node, castClassDepth, castClassDepthReg);
+      generateCompareInstruction(cg, node, instanceClassDepthReg, castClassDepthReg);
+      }
+   srm->reclaimScratchRegister(instanceClassDepthReg);
+
+   // if objectClassDepth is less than or equal to castClassDepth, then call Helper
+   generateConditionalBranchInstruction(cg, TR::InstOpCode::b_cond, node, falseLabel, TR::CC_LE);
+
+   // Load the superclasses array of the instance class and check if the superclass that appears at the depth of the cast class is in fact the cast class.
+   // If not, the instance class and cast class are not in the same hierarchy.
+   //
+   TR::Register *instanceClassSuperClassesArrayReg = srm->findOrCreateScratchRegister();
+   TR::Register *instanceClassSuperClassReg = srm->findOrCreateScratchRegister();
+
+   generateTrg1MemInstruction(cg,TR::InstOpCode::ldrimmx, node, instanceClassSuperClassesArrayReg,
+                              new (cg->trHeapMemory()) TR::MemoryReference(instanceClassReg, offsetof(J9Class, superclasses), cg));
+
+   int32_t castClassDepthOffset = castClassDepth * TR::Compiler->om.sizeofReferenceAddress();
+   if (constantIsUnsignedImm12(castClassDepthOffset))
+      {
+      generateTrg1MemInstruction(cg, TR::InstOpCode::ldrimmx, node, instanceClassSuperClassReg,
+                                 new (cg->trHeapMemory()) TR::MemoryReference(instanceClassSuperClassesArrayReg, castClassDepthOffset, cg));
+      }
+   else
+      {
+      if (!castClassDepthReg)
+         {
+         castClassDepthReg = srm->findOrCreateScratchRegister();
+         loadConstant32(cg, node, castClassDepth, castClassDepthReg);
+         }
+      generateLogicalShiftLeftImmInstruction(cg, node, castClassDepthReg, castClassDepthReg, 3, false);
+      generateTrg1MemInstruction(cg, TR::InstOpCode::ldrimmx, node, instanceClassSuperClassReg, new (cg->trHeapMemory()) TR::MemoryReference(instanceClassSuperClassesArrayReg, castClassDepthReg, cg));
```

Comment:
`ldrimmx` expects a MemoryReference with a base register and a integer offset.
Perhaps you mean `ldroffx`?

Reply:
Thanks. I incorrectly wrote it as `ldrimmx`. I meant `ldroffx`.